src/rust/inetstack/protocols/udp/peer.rs

```diff
@@ -270,93 +309,95 @@ impl UdpPeer {
     pub fn do_close(&mut self, qd: QDesc) -> Result<(), Fail> {
         #[cfg(feature = "profiler")]
         timer!("udp::close");
-
+        let mut qtable: RefMut<IoQueueTable<InetQueue>> = self.qtable.borrow_mut();
         // Lookup associated endpoint.
-        let socket: Option<SocketAddrV4> = match self.sockets.remove(&qd) {
-            Some(s) => s,
-            None => return Err(Fail::new(EBADF, "invalid queue descriptor")),
-        };
-
-        // Remove endpoint binding.
-        match socket {
-            Some(local) if self.bound.remove(&local).is_some() => Ok(()),
-            _ => return Err(Fail::new(EBADF, "invalid queue descriptor")),
+        match qtable.free(&qd) {
+            Some(InetQueue::Udp(queue)) => match queue.get_addr() {
+                Ok(addr) => {
+                    self.bound.remove(&addr);
+                    Ok(())
+                },
+                Err(e) => Err(e),
+            },
+            _ => Err(Fail::new(EBADF, "invalid queue descriptor")),
```

Comment:
Since you are touching this line, could you prefix the error code eith `libc::`?